File: platform/fastapi_server/config/lifecycle/01_app_yaml.lifecycle.py
import logging
from fastapi import FastAPI
from app_yaml_load import load_app_yaml_config

logger = logging.getLogger("lifecycle:app_yaml")


def onInit(app: FastAPI, config: dict):
    """
    Initialize App Yaml Config during onInit phase.
    This runs before onStartup, making config available to other lifecycle hooks.
    """
    logger.info("Starting app_yaml lifecycle hook...")
    try:
        result = load_app_yaml_config()

        app.state.config = result.config
        app.state.sdk = result.sdk

        logger.info("App Yaml Config loaded: %s", app.state.sdk.list_providers())

        # Debug: log security config values
        cors_origins = result.config.get_nested('cors', 'origins', default=[])
        csp_directives = result.config.get_nested('contentSecurityPolicy', 'directives', default={})
        logger.debug("[security.yml] cors.origins: %s", cors_origins)
        for directive, values in csp_directives.items():
            logger.debug("[security.yml] contentSecurityPolicy.directives.%s: %s", directive, values)

        logger.info("app_yaml lifecycle hook completed successfully")
    except Exception as exc:
        logger.error("app_yaml lifecycle hook failed: %s", exc, exc_info=True)
        raise

File: platform/platform-core/fastify/src/lifecycles/01_app_yaml.lifecycle.mjs
import { loadAppYamlConfig } from "@internal/app-yaml-load";

// This file matches the signature bound by `server.mjs`:
// if (module.onStartup) startupHooks.push(module.onStartup);

export async function onStartup(server, config) {
  server.log.info("[lifecycle:app-yaml] Initializing App Yaml Static Config...");

  try {
    const { config: appConfig, sdk } = await loadAppYamlConfig();

    server.log.info({ providers: sdk.listProviders() }, "[lifecycle:app-yaml] Configuration loaded");

    // Debug: print security config values
    const corsOrigins = appConfig.getNested(["cors", "origins"], []);
    const cspDirectives = appConfig.getNested(
      ["contentSecurityPolicy", "directives"],
      {},
    );
    server.log.info({ corsOrigins }, "[lifecycle:app-yaml] [security.yml] cors.origins");
    for (const [directive, values] of Object.entries(cspDirectives)) {
      server.log.debug({ directive, values }, "[lifecycle:app-yaml] [security.yml] contentSecurityPolicy.directives");
    }

    // Decorate Fastify instance for global access.
    // server.mjs and app.mjs intentionally do NOT decorate 'config' so that
    // this hook is the sole owner of server.config (AppYamlConfig instance).
    if (!server.hasDecorator("config")) {
      server.log.info("[lifecycle:app-yaml] Decorating server with AppYamlConfig");
      server.decorate("config", appConfig);
    } else {
      // Fallback: if something else decorated config, force-replace via defineProperty
      server.log.warn("[lifecycle:app-yaml] server.config already decorated — force-replacing with AppYamlConfig");
      Object.defineProperty(server, 'config', {
        value: appConfig,
        writable: true,
        configurable: true,
        enumerable: true,
      });
    }
    if (!server.hasDecorator("sdk")) {
      server.log.info("[lifecycle:app-yaml] Decorating server with sdk");
      server.decorate("sdk", sdk);
    }

    // Verify the decoration worked — catch assignment failures early
    if (typeof server.config?.getNested !== 'function') {
      throw new Error('[lifecycle:app-yaml] CRITICAL: server.config.getNested is not a function after decoration. server.config type: ' + typeof server.config);
    }

    server.log.info("[lifecycle:app-yaml] App Yaml Static Config initialized successfully");
  } catch (err) {
    server.log.error({ err, hookName: "01-app-yaml" }, "[lifecycle:app-yaml] App Yaml Static Config initialization failed");
    throw err;
  }
}
